### DISCUSSION TITLE
How is `args is { Length: > 0 }` better than `args.Length > 0`

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [dotnet/runtime#65385] [Q&A] [answered]
I came across this code, and am a bit confused about why pattern matching is used here. 
Isn't `args.Length > 0` would be better (faster)?

https://github.com/dotnet/runtime/blob/a14d27ac7ff3a58e39d889db72f7ce682c7492fd/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L194 https://github.com/dotnet/runtime/blob/a14d27ac7ff3a58e39d889db72f7ce682c7492fd/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L219

**Comment by stephentoub** [comment, marked as answer]
There's a functional difference.
```C#
if (args is { Length: > 0 }) 
```
is the same as:
```C#
if (args is not null && args.Length > 0)
```
whereas
```C#
if (args.Length > 0)
```
will null ref if `args` is null.  If `args` can't be null, `if (args.Length > 0)` is a tad better.  If it can be null, `if (args is { Length: > 0 })` is clearly better.

(But with the lines you cited, there's no way that null check will have a measurable impact on performance.)

**Reply by maxkoshevoi** [reply]
So that's a null check. Got it, thanks. I'm annotating this area, so that's really helpful!
Btw, do how know whether values in `args` could be `null`?

**Reply by stephentoub** [reply]
> Btw, do how know whether values in args could be null?

From an annotations perspective, it'd be `string?[] args` to indicate it may contain nulls and `string[] args`, but that of course doesn't guarantee it doesn't (e.g. `string[] arr = new string[10]` is an array filled with nulls even though it's annotated as containing non-null strings).  To know whether it actually contains nulls at run-time, you'd need to iterate through the array.

**Reply by maxkoshevoi** [reply]
> it'd be `string?[] args` to indicate it may contain nulls and `string[] args`

Sorry, didn't quite get you. So `string?[] args` or `string[] args`? Interesting point `string[] arr = new string[10]` being filled with `null`s, and compiler not complaining about it, haven't thought about that.

**Reply by stephentoub** [reply]
`string?[] args` is a non-nullable array that might contain nulls.

`string[]? args` is a nullable array that shouldn't contain nulls.

`string[] args` is a non-nullable array that shouldn't contain nulls.

**Reply by maxkoshevoi** [reply]
Yeah, I understand that. I was asking you about that specific case (`IHostBuilder.ConfigureDefaults`).

https://github.com/dotnet/runtime/blob/a14d27ac7ff3a58e39d889db72f7ce682c7492fd/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L188

If you don't know off the top of your head, that's fine. Sorry for misunderstanding

**Reply by stephentoub** [reply]
> I was asking you about that specific case (IHostBuilder.ConfigureDefaults)

I would assume `string[]? args`.  I see callers passing in null, and otherwise the args come from Main, which shouldn't be getting null strings in the array.
